File: app/src/main/java/com/angogasapps/myfamily/objects/Message.java
package com.angogasapps.myfamily.objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

import com.google.firebase.database.DataSnapshot;

@Entity
public class Message implements Comparable<Message>{
    @NonNull
    @PrimaryKey
    private String id = "";

    private String from;
    private String type;
    private long time;
    private String  value;


    public String getFrom() {
        return from;
    }

    public String getType() {
        return type;
    }

    public long getTime() {
        return time;
    }

    public String getValue() {
        return value;
    }

    public String getId(){
        return id;
    }


    @Override
    public boolean equals(@Nullable Object obj) {
        if (!(obj instanceof Message))
            return false;
        return this.id.equals(((Message)obj).getId());
    }

//    public boolean equals(Message obj) {
//        return (this.id.equals(obj.getId()));
//    }

    public Message() {}

    public Message(DataSnapshot snapshot) {
        Message cash = snapshot.getValue(Message.class);
        this.id = snapshot.getKey();
        this.from = cash.getFrom();
        this.type = cash.getType();
        this.time = cash.getTime();
        this.value = cash.getValue();
    }

    @Override
    public int compareTo(Message obj) {
        if (this.time > obj.getTime()) {
            return 1;
        }
        else if (this.time <  obj.getTime()) {
            return -1;
        }
        else {
            return 0;
        }
    }

    public void setId(@NonNull String id) {
        this.id = id;
    }

    public void setFrom(String from) {
        this.from = from;
    }

    public void setType(String type) {
        this.type = type;
    }

    public void setTime(long time) {
        this.time = time;
    }

    public void setValue(String value) {
        this.value = value;
    }
}

File: app/src/main/java/com/angogasapps/myfamily/models/Message.kt
package com.angogasapps.myfamily.models

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.google.firebase.database.DataSnapshot

@Entity
data class Message(
    @PrimaryKey
    var id: String = "",
    var from: String = "",
    var type: String = "",
    var time: Long = 0,
    var value: String = ""
) : Comparable<Message> {

    override fun equals(other: Any?): Boolean {
        return if (other !is Message) false else id == other.id
    }

    constructor() : this(id = "") {}
    constructor(snapshot: DataSnapshot) : this() {
        val cash = snapshot.getValue(
            Message::class.java
        )
        id = snapshot.key!!
        from = cash!!.from
        type = cash.type
        time = cash.time
        value = cash.value
    }

    override fun compareTo(other: Message): Int {
        return when {
            time > other.time -> 1
            time < other.time -> -1
            else -> 0
        }
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + from.hashCode()
        result = 31 * result + type.hashCode()
        result = 31 * result + time.hashCode()
        result = 31 * result + value.hashCode()
        return result
    }
}